### ISSUE TITLE
relax $input in fn:serialize

### ISSUE THREAD

**Issue by Arithmeticus** [qt4cg/qtspecs#321]
Relevant specifications:
https://qt4cg.org/specifications/xpath-functions-40/Overview-diff.html#func-serialize

Would it be possible to relax the strictures on `$input` (first parameter) of `fn:serialize()`?

1. The specifications do not explicitly forbid `map(*)` or `array(*)` as input, but in practice, when these are supplied, Saxon rejects them. Developers (or at least this one) who work with arrays and maps often need to render them in string output or messages, if only for diagnostics. If there is something really prohibitively wrong with those two items as input to `fn:serialize()`, then the specifications should say so.
2. Attributes are forbidden, but it is unclear why. They get serialized fine in the context of a parent, why not alone?
3. Namespace nodes are forbidden; see previous point.

(No doubt there must have been discussion on points 2-3, but the rationale is not clear from the specs.)

Perhaps the question is that the details of what the serialization should look like is contestable. I think the answer there is simply, pick one. I think we'll happily live with whatever is chosen.

For the serialization of maps and arrays, I'll point as one possible model my [tan:map-to-xml()](https://github.com/textalign/TAN-2021/blob/9d2215e36da1f64bddad9daddfc5aa2242e17c81/functions/maps/TAN-fn-maps-extended.xsl#L58) and [tan:array-to-xml()](https://github.com/textalign/TAN-2021/blob/9d2215e36da1f64bddad9daddfc5aa2242e17c81/functions/arrays/TAN-fn-arrays-extended.xsl#L66), which have been indispensable for daily troubleshooting.

**Comment by joewiz** [comment]
Does the adaptive serialization method perhaps meet your need for serializing maps, arrays, and attributes? (Namespace nodes are passed through without error, though not explicitly serialized.) 

Or does the fact that these are not forbidden and do not raise errors suggest that perhaps the language you cited mean that the spec may need to be adjusted to account for the existence of the adaptive serialization method?

**Comment by Arithmeticus** [comment]
Quick question: adaptive serialization method = ?

**Comment by michaelhkay** [comment]
The restrictions on fn:serialize() depend on the output method. You can't serialize a map or array using the XML or HTML output methods because that would be a transformation, not a serialization. Equally, you can't serialize a sequence of element nodes using the JSON serialization method.

The adaptive serialization method was designed as something capable of serializing anything you throw at it, but the output isn't in general going to be in any standard format.

**Comment by ChristianGruen** [comment]
> Quick question: adaptive serialization method = ?

…can e.g. be used as follows:

```xquery
serialize(
  map { 'key': [ 'value1', ('value2', 'value3') ] },
  map { 'method': 'adaptive' }
)
```

Saxon and BaseX both return the following result…

```xquery
map{"key":["value1",("value2","value3")]}
```

…but as Michael indicated, many details are implementation-defined and may change over time.

Details can be found in the [XSLT and XQuery Serialization Specification](https://www.w3.org/TR/xslt-xquery-serialization-31/#adaptive-output).

**Comment by Arithmeticus** [comment]
Thanks, this is a case where relying on the FO specs as my point of entry left me unaware of the adaptive method, and probably of a number of other useful features to `fn:serialize()`. (Sorry, but a stand-off document the size and verbal density of the Serialize specs, not one but two clicks away, is not helpful for quickly understanding core features of `fn:serialize()`.) Some minor edits in the FO specs would help. The specs currently have notes that target three different use cases. I recommend adding another note that highlights the use case of maps and arrays, explicitly citing `map { 'method': 'adaptive' }`. Maybe add more examples.

In the case of attributes and namespace nodes, they are specified without exception as error-triggering. Maybe the language should be qualified (current text roman; suggested insertion italics): "The serialization process will raise an error if $input is an attribute or namespace node _, unless `method` is set to `adaptive` and the implementation supports that serialization method_."

**Comment by pgfearo** [comment]
Perhaps there's a case for a standard XML representation of the XPath Data Model, effectively an extension to what is used for JSON in XSLT 3.0? That is, with elements like 'map', 'array', 'sequence' etc

**Comment by line-o** [comment]
Just for completeness exist-db also does support adaptive serialization. The above query yields:
```
map{"key":["value1",("value2","value3"),fn:true#0,(anonymous-function)#0]}
```

**Comment by ChristianGruen** [comment]
> Just for completeness, exist-db also does support the adaptive serialization method. The above query yields:

Good to know, thanks; and even better to see that the output is identical.

**Comment by line-o** [comment]
I was relieved, too :)

**Comment by line-o** [comment]
Since it was missing in the example, I was curious if the adaptive output for attribute nodes would differ.

```xquery
serialize(attribute test { 'the value' },map{'method':'adaptive'})
```

yields `test="the value"` in both basex and exist

**Comment by ndw** [comment]
Resolved by #472
